Mathlib/Logic/Function/OfHArity.lean

```diff
@@ -0,0 +1,101 @@
+Authors: Brendan Murphy
+-/
+
+import Mathlib.Logic.Function.OfArity
+import Mathlib.Data.Fin.VecNotation
+
+/-! # Function types of a given heterogeneous arity
+
+This provides `FunctionOfHArity`, such that `OfHArity ![α, β] τ = α → β → τ`.
+Note that it is often preferable to use `((i : Fin n) → p i) → τ` in place of `OfArity p τ`.
+
+## Main definitions
+
+* `Function.OfHArity p τ`: `n`-ary function `p 0 → p 1 → ... → p (n - 1) → β`.
+-/
+
+universe u
+
+namespace Function
+
+open Matrix (vecCons vecHead vecTail vecEmpty)
+
+/-- The type of `n`-ary functions `p 0 → p 1 → ... → p (n - 1) → τ`. -/
+def OfHArity : {n : ℕ} → (Fin n → Type u) → Type u → Type u
+  | 0    , _, τ => τ
+  | n + 1, p, τ => vecHead p → @OfHArity n (vecTail p) τ
+
+@[simp]
+theorem ofHArity_zero (p : Fin 0 → Type u) (τ : Type u) : OfHArity p τ = τ := rfl
+
+theorem ofHArity_nil (τ : Type u) : OfHArity ![] τ = τ := ofHArity_zero ![] τ
+
+@[simp low] -- prefer `ofHArity_cons` when it (syntactically) applies
+theorem ofHArity_succ {n} (p : Fin (n + 1) → Type u) (τ : Type u) :
+    OfHArity p τ = (vecHead p → OfHArity (vecTail p) τ) := rfl
+
+@[simp]
+theorem ofHArity_cons {n} (α : Type u) (p : Fin n → Type u) (τ : Type u) :
+    OfHArity (vecCons α p) τ = (α → OfHArity p τ) := ofHArity_succ _ τ
+
+/-- The (identity) equivalence between 0-ary heterogeneous functions and
+the terminal codomain type. -/
+def ofHArity_zero_equiv (p : Fin 0 → Type u) (τ : Type u) :
+    OfHArity p τ ≃ τ := Equiv.refl _
+
+/-- The (identity) equivalence between `![]`-ary heterogeneous functions and
+the terminal codomain type. -/
+def ofHArity_nil_equiv (τ : Type u) : OfHArity ![] τ ≃ τ :=
+  ofHArity_zero_equiv ![] τ
+
+/-- The (identity) equivalence between `p`-ary heterogeneous functions and
+unary functions from the head of `p` to `vecTail p`-ary heterogeneous functions. -/
+def ofHArity_succ_equiv {n} (p : Fin (n + 1) → Type u) (τ : Type u) :
+    OfHArity p τ ≃ (vecHead p → OfHArity (vecTail p) τ) := Equiv.refl _
+
+/-- The (identity) equivalence between `(vecCons α p)`-ary heterogeneous functions and
+unary functions from `α` to `p`-ary heterogeneous functions. -/
+def ofHArity_cons_equiv {n} (α : Type u) (p : Fin n → Type u) (τ : Type u) :
+    OfHArity (vecCons α p) τ ≃ (α → OfHArity p τ) := ofHArity_succ_equiv _ _
+
+-- not a definitional equality, generally prefer `ofHArity_fin_const_equiv`
+lemma ofHArity_fin_const {n} (α β : Type u) :
+    OfHArity (fun (_ : Fin n) => α) β = OfArity α β n :=
+  match n with
+  | 0   => Eq.refl β
+  | n+1 => congrArg (α → .) (@ofHArity_fin_const n α β)
+
+/-- The equivalence (not a definitional equality!) between `n`-ary heterogeneous
+functions that aren't actually heterogeneous and `n`-ary functions. -/
+def ofHArity_fin_const_equiv : {n : ℕ} → (α β : Type u) →
+    OfHArity (fun (_ : Fin n) => α) β ≃ OfArity α β n
+  | 0,   _, β => .refl β
+  | n+1, α, β => .arrowCongr (.refl α) (@ofHArity_fin_const_equiv n α β)
```

Comment:
Is it possible to redefine `OfArity` such that this _is_ a definitional equality? Or does something break?

Reply:
That's definitely possible, and I'm happy to make that change. I was just leery of modifying existing code

Reply:
I think it would make a nice standalone PR to introduce `OfHArity`, and use it to replace most of the `OfArity` API.

Reply:
Good idea, I'm going to close this now and open a standalone PR for that (and in time try to PR in the ternary currying for functors stuff)